src/posts/interning-at-khan-academy.md

```diff
@@ -0,0 +1,55 @@
+
+![changes to the math keypad](/images/interning-at-khan-academy/image_0.png)
+
+*an early screenshot of changes to the math keypad*
+
+# Part 2: iframes, iframes, iframes
+
+Now that I was used to the flow of picking a task, creating a code review, and deploying my changes, I started on my first project: improving our content editor to be mobile first. Historically, our content creators have worked primarily on desktop: they write their content on desktop, they preview their content on desktop, they review each other's work on desktop, they check their email on desktop. How can we make it easy for our content creators to build content for mobile first, and thereby avoid awful small-screen layout bugs, but still allow them to take advantage of the big-screen multi-tasking perks of desktop devices? This involved pulling together specs for what types of devices students use Khan Academy on, rendering content with the new styles the rest of the team was working on, isolating the rendered content to prevent styles leaking from the rest of the editor, and implementing touch emulation for interacting with the mobile previews.
+
+To implement the isolation of content, I decided to render previews inside of iframes. In addition to providing style isolation, iframes also made it possible to properly render components that are rendered directly to the body, such as the math keypad I worked on earlier. But to work with iframes, I now had to make sure that the iframes showed live previews while maintaining the speed of the preview updates. Having quick preview updates meant that iframe reloads had to be minimized and communication between the editor and preview frames needed to be as fast as possible. To transfer data between the parent editor and the child iframe, the parent first stores a JS object representation of the exercise content in a global variable, because the exercise content contains callback functions which prevent it from being serialized for window messaging. The parent then sends a message ([http://ejohn.org/blog/cross-window-messaging/](http://ejohn.org/blog/cross-window-messaging/)) to the iframe, notifying it of the updated content.
+
+When the iframe receives this message, it grabs the global JS object from the parent (because the iframe and parent are on the same domain this is allowed by the browser) and renders that content. With this implementation, preview updates happen immediately after content edits. In addition, with an optimized `shouldComponentUpdate` that only allows iframe reloads when they are needed for changes such as touch emulation, iframe loads are minimized so that in most cases the only loading delay happens during the initial editor load.
```

Comment:
"it grabs the global JS object" => "it grabs the exercise content from the global variable".

I think this will read better if you extract the stuff inside the parenthesis into it's own sentence.  "...and render that content.  The browser only allows this because the iframe and parent are on the same domain."

This last sentence could be broken up to read a little better:
"In addition, iframe loads are minimized with a custom `shouldComponentUpdate` method that only allows full reloads for switching form factor or enabling/disabling touch emulation.  In most cases, the only delay happens during the initial editor load."